src/main/scala/viper/gobra/frontend/info/implementation/property/DependencyAnalysis.scala

```diff
@@ -0,0 +1,130 @@
+    */
+  def globalsExprDependsOn(exp: PExpression): Either[String, Vector[st.GlobalVariable]] = {
+    val enclosingPkg = tryEnclosingPackage(exp).get
+    val stack = mutable.Stack[PNode](exp)
+    var visitedEntities = Seq[st.Regular]()
+    // Stores all dependencies of a node (including itself).
+    var dependencies = Seq[st.GlobalVariable]()
+    var foundError: Option[String] = None
+
+    // According to the Go language spec:
+    //  Dependency analysis does not rely on the actual values of the variables, only on lexical
+    //  references to them in the source, analyzed transitively.
+    //  For instance, if a variable x's initialization expression refers to a function whose body
+    //  refers to variable y then x depends on y. Specifically:
+    //    - A reference to a variable or function is an identifier denoting that variable or function.
+    //    - A reference to a method m is a method value or method expression of the form t.m, where the (static)
+    //      type of t is not an interface type, and the method m is in the method set of t. It is immaterial
+    //      whether the resulting function value t.m is invoked.
+    //    - A variable, function, or method x depends on a variable y if x's initialization expression or
+    //      body (for functions and methods) contains a reference to y or to a function or
+    //      method that depends on y.
+    while (stack.nonEmpty && foundError.isEmpty) {
+      val elem = stack.pop()
+      reflexiveAllChildren(elem).foreach{
+        case n: PNamedOperand =>
+          // If it is a function call, it is not a ghost call,
+          // and it is not dynamically-bound, collect the method info if it has not been collected.
+          // If it is a global variable, collect its dependencies if it has not been traversed yet.
+          resolve(n) match {
+            case Some(g: ap.GlobalVariable) if !isEnclosingGhost(n) =>
+              if (!visitedEntities.contains(g.symb)) {
+                dependencies :+= g.symb
+                visitedEntities :+= g.symb
+                stack.push(g.symb.decl)
+              }
+            case Some(f: ap.Function) if !isEnclosingGhost(n) =>
+              if (!visitedEntities.contains(f.symb)) {
+                visitedEntities :+= f.symb
+                stack.push(f.symb.decl)
+              }
+            case _ =>
+          }
+        case n: PDot =>
+          // If it is from another package, do not analyse.
+          // If it is a reference to a method declared in the current package, it is not in a ghost-context,
+          // and it is not dynamically-bound, collect the method info if it has not been collected.
+          resolve(n) match {
+            case Some(f: ap.ReceivedMethod) if !isEnclosingGhost(n) &&
+              underlyingType(typ(f.recv)).isInstanceOf[Type.InterfaceT] =>
+              foundError = Some(s"Calls to dynamically-bound non-ghost methods are not allowed in initialization code, but found $n.")
```

Comment:
these intendations make it hard to distinguish what is the condition and what is the code.

Reply:
That is true, but at the same time, the choice is between this and making the lines kind of long. I went with the latter now, given that it looks more readable to me